sw/device/lib/testing/mask_rom_test.h

```diff
@@ -0,0 +1,93 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_TESTING_MASK_ROM_TEST_H_
+#define OPENTITAN_SW_DEVICE_LIB_TESTING_MASK_ROM_TEST_H_
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace mask_rom_test {
+
+/**
+ * Mixin for global mocks.
+ *
+ * `GlobalMock<T>` is a subclass of `T` that can have at most one instance at a
+ * time (checked at runtime) and makes this instance globally accessible via the
+ * static `GlobalMock<T> &Instance()` method.
+ *
+ * Mock classes used in tests should be of this type so that mock functions can
+ * access their instances during tests. Since we prefer using strict mocks,
+ * mock classes should also be of type `testing::StrictMock`. Mock classes that
+ * satisfy both requirements can be defined using a type alias as follows:
+ *
+ *     namespace mask_rom_test {
+ *     namespace internal {
+ *     class MockFoo {
+ *       ...
+ *     };
+ *     }  // namespace internal
+ *     // Type alias for making `internal::MockFoo` a global and strict mock.
+ *     using MockFoo = GlobalMock<testing::StrictMock<internal::MockFoo>>;
+ *     ...
+ *     }  // namespace mask_rom_test
+ */
+template <typename T>
+class GlobalMock : public T {
+ public:
+  GlobalMock() {
+    if (instance_ != nullptr) {
+      throw std::runtime_error("Mock is already instantiated.");
+    }
+    instance_ = this;
+  }
+
+  virtual ~GlobalMock() { instance_ = nullptr; }
+
+  static GlobalMock<T> &Instance() {
+    if (instance_ == nullptr) {
+      throw std::runtime_error("Mock is not instantiated yet.");
+    }
+    return *instance_;
+  }
+
+ protected:
+  GlobalMock(const GlobalMock &) = delete;
+  GlobalMock &operator=(const GlobalMock &) = delete;
+  GlobalMock(GlobalMock &&) = delete;
+  GlobalMock &operator=(GlobalMock &&) = delete;
+
+ private:
+  static GlobalMock<T> *instance_;
+};
+template <typename T>
+GlobalMock<T> *GlobalMock<T>::instance_ = nullptr;
+
+/**
+ * Test fixture for mask ROM tests.
+ *
+ * Test suites should derive their test fixtures from this class. This class
+ * enforces that mock methods are called in the order `EXPECT_CALL` statements
+ * are written. In cases where this behavior is not desired, test fixtures can
+ * derive from `Unordered<MaskRomTest>` instead to opt-out.
+ */
+class MaskRomTest : public testing::Test {
+ protected:
+  std::unique_ptr<testing::InSequence> seq_ =
+      std::make_unique<testing::InSequence>();
+};
+
+/**
+ * Mixin for unordered calls.
+ *
+ * @see MaskRomTest.
+ */
+template <typename T>
+class Unordered : public T {
+ protected:
+  Unordered() : T() { T::seq_.reset(); }
+};
```

Comment:
I'm still not sure if I'm happy with this default. It feels like testing for function calls doesn't need to be by-default in sequence like MMIO calls are, and I feel it would make the tests more consistent with usual googletest style to be out-of-order by default.

I'm mostly trying to stick to googletest convention, but I understand if we can't in some places because of our slightly weird use of it.

(Separately, I think that the `Unordered<MaskRomTest>` idiom is slightly bizarre, since `Unordered<T>` is meaningless except for `T = MaskRomTest`; at that point, you might as well not use MaskRomTest at all...)

Reply:
I understand, but we can always revisit this and I think this is a reasonable place to start. Let's keep an eye on the tests and see which case is used most. Regarding the mixin, yes, it feels pointless since `MaskRomTest` is such a thin class but I suggest we move forward and see if need it at all.